### PR TITLE
Add bot service

### PR DESCRIPTION
**Pull request by Schmavery**
Initial commit of reason-bot-service.
Entrypoint is `App.re`.
Should be relatively straightforward to follow starting from there.
The code calls out to the Faucet.handleMessage for faucet stuff, and Echo.start() starts an echo service.

- Manually tested with a test discord bot on the current testnet
- Added CI to make sure it builds

### CODE REVIEW

File: frontend/bot/package.json [hunk 1]
```diff
@@ -0,0 +1,24 @@
+{
+  "name": "coda-bot",
+  "version": "1.0.0",
+  "main": "index.js",
+  "license": "MIT",
```
Comment (wu-s-john) [inline frontend/bot/package.json]: our license should be Apache 2 

Ref: @cmr

Comment (Schmavery) [reply frontend/bot/package.json]: thx, fixed (I'll make another PR to fix it in the other package.jsons)

File: frontend/bot/src/App.re [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  | ["$request", pk] => Faucet.handleMessage(msg, pk)
+  | ["$request", ..._] => Message.reply(msg, Messages.requestError)
+  | _ => ()
+  };
+
+let handleMessage = msg =>
+  switch (TextChannel.fromChannel(Message.channel(msg))) {
+  | None => ()
+  | Some(channel) =>
+    if (!(
```
Comment (o1pranay) [inline frontend/bot/src/App.re]: I think this handles the message if user is a bot and channel is not in the list?

should it be: `if (!bot && correctChannel)` instead?

File: frontend/bot/src/App.re [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  };
+
+// Start echo service
+Echo.start(Constants.echoKey);
+
+Client.onReady(client, _ => print_endline("Bot is ready"));
+
+Client.onMessage(client, handleMessage);
+
+Client.login(client, "TOKEN_HERE");
```
Comment (o1pranay) [inline frontend/bot/src/App.re]: sir where token?

File: frontend/bot/src/Coda.re [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    publicKey @bsDecoder(fn: "Graphql.Decoders.publicKey")
+  }} |}
+];
+
+let printWallets = () => {
+  Graphql.executeQuery(Wallets.make())
+  |> Wonka.forEach((. response) =>
+       switch (response) {
+       | Graphql.Data(data) =>
+         print_endline("Wallets:");
```
[the remark this answers is not in the document]
Comment (Schmavery) [reply frontend/bot/src/Coda.re]: This code doesn't run, I could just delete it if you prefer

File: frontend/bot/src/App.re (hunk 2 above)
Comment (Schmavery) [reply frontend/bot/src/App.re]: good call

File: frontend/bot/src/Messages.re [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+  Woof! I can help you get some coda on the testnet.
+Just send me a message in the #faucet channel with the following contents:
+`$request <public-key>`
+Once a mod approves, `|}
+  ++ Int64.to_string(Constants.faucetAmount)
+  ++ {| coda` will be sent to the requested address!|};
+
+let requestError = {|
+Grrrrr... Invalid Parameters!!
+`$request <public-key>`|};
```
[the remark this answers is not in the document]
Comment (Schmavery) [reply frontend/bot/src/Messages.re]: Sure, I was just copying the existing code but that makes sense

File: frontend/bot/src/App.re (hunk 2 above)
Comment (o1pranay) [reply frontend/bot/src/App.re]: no worries :D

File: frontend/bot/src/Graphql.re [hunk 6]
```diff
@@ -0,0 +1,108 @@
+%raw
+"global.fetch = require(\"node-fetch\");";
```
Comment (bkase) [inline frontend/bot/src/Graphql.re]: lol

File: frontend/bot/src/Graphql.re [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+[@bs.deriving abstract]
+type responseJs = {
+  data: Js.Nullable.t(Js.Json.t),
+  [@bs.optional]
+  error: err,
+};
+
+let processResponse = (parse, response) => {
+  let data = response->dataGet->Js.Nullable.toOption->Belt.Option.map(parse);
+  let error = response->errorGet;
```
Comment (bkase) [inline frontend/bot/src/Graphql.re]: I assume you just pasted this from somewhere, but mind switching to `|>` and not using belt

Comment (bkase) [reply frontend/bot/src/Graphql.re]: (tablecloth I guess?)

Comment (Schmavery) [reply frontend/bot/src/Graphql.re]: 🙄

**Comment by bkase** [comment]
We can try belt on this project

**Pull request merged**